cvat/apps/dataset_manager/bindings.py

```diff
@@ -2151,11 +2151,89 @@ def reduce_fn(acc, v):
                 raise CvatImportError("Image {}: can't import annotation "
                     "#{} ({}): {}".format(item.id, idx, ann.type.name, e)) from e
 
-    for track in tracks.values():
+    stop_frame = int(instance_data.meta[instance_data.META_FIELD]['stop_frame'])
+    for track_id, track in tracks.items():
+        # skip tracks with no shapes and elements
+        if len(track['shapes']) == 0 and len(track['elements']) == 0:
+            continue
+
+        track['shapes'].sort(key=lambda t: t.frame)
+        prev_shape = track['shapes'][0]
+        # add keyframe to the first non-outside shape
+        if not prev_shape.outside:
+            prev_shape = prev_shape._replace(keyframe=True)
+        new_shapes = []
+
+        # avoid skipping the first frame
+        if len(track['shapes']) > 1:
+            for shape in track['shapes'][1:]:
+                if prev_shape.keyframe or prev_shape.outside:
```

Comment:
Probably, the shape has only be added if it's visible. Actually, I recommend to exclude all the outside shapes to simplify the code. This way, you'll only need to check if the previous was `None` or there was a skip.